### PR TITLE
PIM-10545 - Remove merge of non viewable categories on SetCategories userIntent

### PR DESCRIPTION
**Pull request by cemorin**
**Description (for Contributor and Core Developer)**

- Remove merge of non viewable categories in user intent
- Check that the user passed in the UpsertProductCommand is the same as the one connected

Before, we merged non viewable categories with the one passed in the userIntent.
But there was a check afterwards on the GrantedProductUpdater for categories passed in the update.

The merge is already done in `NotGrantedCategoryMerger`

**Definition Of Done (for Core Developer only)**

- [x] Tests
- [ ] Migration & Installer
- [ ] PM Validation (Story)
- [ ] Changelog (maintenance bug fixes)
- [ ] Tech Doc

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php [hunk 1]
```diff
@@ -33,7 +37,8 @@ public function __construct(
         private SaverInterface $productSaver,
         private ValidatorInterface $productValidator,
         private EventDispatcherInterface $eventDispatcher,
-        private UserIntentApplierRegistry $applierRegistry
+        private UserIntentApplierRegistry $applierRegistry,
+        private UserContext $userContext,
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php:41]:
```suggestion
        private UserContext $userContext,
```

Whydo we do that? UserContext is a kind of anti pattern and we should avoid its usage.

Comment (cemorin) [reply src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php:41]: We want to make sure the userId provided to the method is the same as the authenticatedUser.
Is it better to use this `$this->tokenStorage->getToken()->getUser()->getId()` ?

Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php:41]: > We want to make sure the userId provided to the method 
But why?

The user is passed by parameters, it's the responsibility of the caller to provide the user.

Comment (cemorin) [reply src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php]: As seen together, the code has been extracted with a more detailed explaination
https://github.com/akeneo/pim-community-dev/pull/17482/files#diff-9dc640bb42bdc456c174dae080c8a581ae2b8df54ac5e8ebbc4421865816f7a2R130

File: src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php [hunk 2]
```diff
@@ -121,4 +125,18 @@ private function updateProduct(ProductInterface $product, UpsertProductCommand $
             }
         }
     }
+
+    /**
+     * Legacy update uses context to check permissions on the current connected user rather than using userId passed
```
Comment (ahocquard) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php:130]:
```suggestion
     * The new service API is for now a wrapper of the legacy implementation to update a product. The legacy implementation use in some services the user context (mainly to check permissions). That's why, for now, we check that the user context is consistent with the provided user id.
     *
     * Once we re-implement the legacy implementation to not rely on the user context, this check and the injection of the user context should be removed. Indeed, for the user context make this service API stateful and not agnostic of the user.
```

Comment (ahocquard) [reply src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php:130]: My advice for comments:
- the why, not the what (it is the case in your comment :+1: )
- if it is a workaround (lacking in your comment)
- should it be removed later on (lacking in your comment)

And if in 1 year, you will understand yourself.  :smile:

File: src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php [hunk 3]
```diff
@@ -121,4 +127,22 @@ private function updateProduct(ProductInterface $product, UpsertProductCommand $
...
+
+    /**
+     * The new service API is for now a wrapper of the legacy implementation to update a product. The legacy
+     * implementation use in some services the user context and not the user id of the command (mainly to check permissions).
+     * That's why, for now, we check that the user context is consistent with the provided user id.
+     *
+     * Once we re-implement the legacy implementation to not rely on the user context, this check and the injection of
+     * the user context should be removed. Indeed, for the user context make this service API stateful and not agnostic of the user.
+     */
+    private function checkConsistencyWithConnectedUser(int $userId): void
```
Comment (RiffVentura) [inline src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php:139]: This adds a requirement to service API that  to use `UpsertProductCommand`, `$userId` provided must be connected.
Is this change intended?

Comment (cemorin) [reply src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php:139]: Yes it is intended, the legacy update service uses the connected user to check on permissions (and does not use the provided userId yet).
So we added this check to make sure there are no inconsistencies between the userId passed to the command, and the following checks that will be done with the connected user (for exemple in `NotGrantedCategoryMerger.php`)
If the explaination above the function is not clear enough let us know what we can add :+1:

Comment (RiffVentura) [reply src/Akeneo/Pim/Enrichment/Product/back/Application/UpsertProductHandler.php:139]: The explanation is clear but my point was more about the service API change and its users (the red catalogs CI step :sweat_smile: ).  Just making sure :wink:

File: src/Akeneo/Pim/Enrichment/Product/back/Test/.php_cd.php [hunk 4]
```diff
@@ -54,7 +54,10 @@
 
         // Public APIs
         'Akeneo\Pim\Structure\Component\AttributeTypes',
-        'Akeneo\Pim\Structure\Component\Repository\AttributeRepositoryInterface'
+        'Akeneo\Pim\Structure\Component\Repository\AttributeRepositoryInterface',
+
+        'Akeneo\UserManagement\Component\Model\UserInterface',
```
Comment (cemorin) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/.php_cd.php:59]: Add comment to say when to remove this

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php [hunk 5]
```diff
@@ -1510,4 +1512,17 @@ private function createRecords(string $referenceEntityCode, array $recordCodes):
...
+        $session = $this->get('session');
+        $user = $this->get('pim_user.repository.user')->findOneByIdentifier($username);
+        Assert::assertNotNull($user);
+
+        $token = new UsernamePasswordToken($user, null, 'main', $user->getRoles());
+        $this->get('security.token_storage')->setToken($token);
+
+        $session->set('_security_main', serialize($token));
+        $session->save();
+    }
```
[the remark this answers is not in the document]
Comment (cemorin) [reply src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductIntegration.php]: Done :)

File: src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php [hunk 6]
```diff
@@ -24,6 +24,8 @@
 use Akeneo\Test\Pim\Enrichment\Product\Helper\FeatureHelper;
 use Akeneo\Test\Pim\Enrichment\Product\Integration\EnrichmentProductTestCase;
 use PHPUnit\Framework\Assert;
+use Symfony\Component\BrowserKit\Cookie;
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Product/back/Test/Integration/Handler/UpsertProductWithPermissionIntegration.php:27]: useless imports

**Review by pierallard — APPROVED** [review]

**Review by Balamung — APPROVED** [review]

**Review by RiffVentura — APPROVED** [review]
